OrbitSsh/Channel.cpp

```diff
@@ -11,195 +11,139 @@
-  }
-}
-
-ResultType Channel::CreateOpenSession(Session* session_ptr,
-                                      std::optional<Channel>* channel_opt) {
+outcome::result<Channel> Channel::OpenChannel(Session* session) {
   LIBSSH2_CHANNEL* raw_channel_ptr =
-      libssh2_channel_open_session(session_ptr->GetRawSessionPtr());
+      libssh2_channel_open_session(session->GetRawSessionPtr());
   if (raw_channel_ptr != nullptr) {
-    *channel_opt = Channel(raw_channel_ptr);
-    return ResultType::kSuccess;
+    return outcome::success(Channel{raw_channel_ptr});
   }
 
-  int last_errno = libssh2_session_last_errno(session_ptr->GetRawSessionPtr());
-  if (last_errno == LIBSSH2_ERROR_EAGAIN) return ResultType::kAgain;
-
-  ERROR("Unable to open channel, last errorno %d", last_errno);
-  return ResultType::kError;
+  return static_cast<Error>(
+      libssh2_session_last_errno(session->GetRawSessionPtr()));
 }
 
-ResultType Channel::CreateDirectTcpIp(Session* session_ptr,
-                                      const std::string& third_party_host,
-                                      int third_party_port,
-                                      std::optional<Channel>* channel_opt) {
-  LIBSSH2_CHANNEL* raw_channel_ptr =
-      libssh2_channel_direct_tcpip(session_ptr->GetRawSessionPtr(),
-                                   third_party_host.c_str(), third_party_port);
+outcome::result<Channel> Channel::OpenTcpIpTunnel(
+    Session* session, const std::string& third_party_host,
+    int third_party_port) {
+  LIBSSH2_CHANNEL* raw_channel_ptr = libssh2_channel_direct_tcpip(
+      session->GetRawSessionPtr(), third_party_host.c_str(), third_party_port);
   if (raw_channel_ptr != nullptr) {
-    *channel_opt = Channel(raw_channel_ptr);
-    return ResultType::kSuccess;
+    return outcome::success(Channel{raw_channel_ptr});
   }
 
-  int last_errno = libssh2_session_last_errno(session_ptr->GetRawSessionPtr());
-  if (last_errno == LIBSSH2_ERROR_EAGAIN) return ResultType::kAgain;
-
-  // TODO (antonrohr) better error handling. (resolve that in higher level)
-  if (last_errno == LIBSSH2_ERROR_CHANNEL_FAILURE) return ResultType::kAgain;
-
-  ERROR("Unable to open channel, last errno %d", last_errno);
-  return ResultType::kError;
+  return static_cast<Error>(
+      libssh2_session_last_errno(session->GetRawSessionPtr()));
 }
 
-ResultType Channel::Exec(const std::string& command) {
-  int rc = libssh2_channel_exec(raw_channel_ptr_, command.c_str());
-  if (rc == 0) return ResultType::kSuccess;
-
-  if (rc == LIBSSH2_ERROR_EAGAIN) return ResultType::kAgain;
+outcome::result<void> Channel::Exec(const std::string& command) {
+  const int rc = libssh2_channel_exec(raw_channel_ptr_.get(), command.c_str());
+  if (rc == 0) {
+    return outcome::success();
+  }
 
-  ERROR("Unable to perform channel exec, error code %d", rc);
-  return ResultType::kError;
+  return static_cast<Error>(rc);
 }
 
-ResultType Channel::Read(std::string* result, int buffer_size) {
```

Comment:
same as with Socket. Does not read everything anymore, just maximum of buffer_size (not in chunks buffer_size). We have to remember that in higher level

Reply:
This should also be taken care of with the check in Tunnel Manager. At least when you don't use a raw channel. Then the user has to take care of it which is IMHO fine.